server/src/main/java/org/opensearch/gateway/PrimaryShardAllocator.java

```diff
@@ -110,17 +112,55 @@ public AllocateUnassignedDecision makeAllocationDecision(
             }
             return AllocateUnassignedDecision.no(UnassignedInfo.AllocationStatus.FETCHING_SHARD_DATA, nodeDecisions);
         }
+        return null;
+    }
 
+    @Override
+    public AllocateUnassignedDecision makeAllocationDecision(
+        final ShardRouting unassignedShard,
+        final RoutingAllocation allocation,
+        final Logger logger
+    ) {
+        AllocateUnassignedDecision decision = getInEligibleShardDecision(unassignedShard, allocation);
+        if (decision != null) {
+            return decision;
+        }
         final FetchResult<NodeGatewayStartedShards> shardState = fetchData(unassignedShard, allocation);
         if (shardState.hasData() == false) {
             allocation.setHasPendingAsyncFetch();
             List<NodeAllocationResult> nodeDecisions = null;
-            if (explain) {
+            if (allocation.debugDecision()) {
                 nodeDecisions = buildDecisionsForAllNodes(unassignedShard, allocation);
             }
             return AllocateUnassignedDecision.no(AllocationStatus.FETCHING_SHARD_DATA, nodeDecisions);
         }
+        List<NodeShardState> nodeShardStates = adaptToNodeShardStates(shardState);
+        return getAllocationDecision(unassignedShard, allocation, nodeShardStates, logger);
+    }
 
+    private static List<NodeShardState> adaptToNodeShardStates(FetchResult<NodeGatewayStartedShards> shardsState) {
+        List<NodeShardState> nodeShardStates = new ArrayList();
+        shardsState.getData().forEach((node, nodeGatewayStartedShard) -> {
+            nodeShardStates.add(
+                new NodeShardState(
+                    node,
+                    nodeGatewayStartedShard.allocationId(),
+                    nodeGatewayStartedShard.primary(),
+                    nodeGatewayStartedShard.replicationCheckpoint(),
+                    nodeGatewayStartedShard.storeException()
+                )
+            );
+        });
+        return nodeShardStates;
+    }
+
+    protected AllocateUnassignedDecision getAllocationDecision(
```

Comment:
We can avoid making any change in the content of this class by reusing the old `NodeGatewayStartedShard` and we need to also include the `hasData` check and remove this check from child  class. Basically, `hasData` also moves to a common place.

Reply:
Addressed in latest commits